src/OpenDebugAD7/TracepointManager.cs

```diff
@@ -0,0 +1,229 @@
+        {
+            string logMessage;
+            if (!this.TryGetValue(breakpointId, out logMessage))
+            {
+                return string.Empty;
+            }
+
+            // There is strings to interpolate in the log message.
+            if (regex.IsMatch(logMessage))
+            {
+                logMessage = GetInterpolatedLogMessage(logMessage, pThread, radix);
+            }
+
+            return logMessage;
+        }
+
+        private string GetInterpolatedLogMessage(string logMessage, IDebugThread2 pThread, uint radix)
+        {
+            if (pThread == null)
+            {
+                return logMessage;
+            }
+
+            // Get topFrame
+            IEnumDebugFrameInfo2 frameInfoEnum;
+            int hr = pThread.EnumFrameInfo(enum_FRAMEINFO_FLAGS.FIF_FRAME | enum_FRAMEINFO_FLAGS.FIF_FLAGS, Constants.EvaluationRadix, out frameInfoEnum);
+            if (hr < 0)
+            {
+                return logMessage;
+            }
+
+            FRAMEINFO[] topFrame = new FRAMEINFO[1];
+            uint fetched = 0;
+            hr = frameInfoEnum.Next(1, topFrame, ref fetched);
+            if (hr < 0 || fetched != 1 || topFrame[0].m_pFrame == null)
+            {
+                return logMessage;
+            }
+
+            HashSet<string> tokensUsed = new HashSet<string>();
+            string interpolatedLogMessage = regex.Replace(logMessage, (match) =>
+            {
+                if (match.Success && !string.IsNullOrWhiteSpace(match.Value))
+                {
+                    char c = match.Value[0];
+                    if (c == '$')
+                    {
+                        string token = match.Value.Substring(1);
+                        switch (token)
+                        {
+                            case "FILEPOS":
+                            case "FUNCTION":
+                            case "ADDRESS":
+                            case "TID":
+                            case "TNAME":
+                            case "PID":
+                            case "PNAME":
+                            case "CALLER":
+                            case "CALLSTACK":
+                            case "TICK":
+                                tokensUsed.Add(token);
+                                return InterpolateToken(token);
+                            default:
+                                return match.Value;
+                        }
+                    }
+                    else
+                    {
+                        string expression = match.Value.Substring(1, match.Value.Length - 2);
+                        try
+                        {
+                            return InterpolateVariable(expression, topFrame[0].m_pFrame, radix);
+                        }
+                        catch (AD7Exception e)
+                        {
+                            return e.ToString();
```

Comment:
>e.ToString(); [](start = 35, length = 13)

Are there cases where this is useful? Sholud we just have some sort of generic 'Unable to exvaluate ...' instead?